test/util/utils.go

```diff
@@ -0,0 +1,394 @@
+		return "", err
+	}
+
+	tdHash := fmt.Sprintf("%x", md5.Sum(tdData))
+	idempotentFamily := *registerRequest.Family + "-" + tdHash
+
+	existing, err := ECS.DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
+		TaskDefinition: &idempotentFamily,
+	})
+	if err == nil {
+		return fmt.Sprintf("%s:%d", *existing.TaskDefinition.Family, *existing.TaskDefinition.Revision), nil
+	}
+
+	registerRequest.Family = &idempotentFamily
+
+	registered, err := ECS.RegisterTaskDefinition(registerRequest)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s:%d", *registered.TaskDefinition.Family, *registered.TaskDefinition.Revision), nil
+}
+
+type TestAgent struct {
+	Image                string
+	DockerID             string
+	IntrospectionURL     string
+	Version              string
+	ContainerInstanceArn string
+	Cluster              string
+	TestDir              string
+
+	DockerClient *docker.Client
+	t            *testing.T
+}
+
+// RunAgent launches the agent and returns an object which may be used to reference it.
+// It will wait until the agent is correctly registered before returning.
+// 'version' may be a docker image (e.g. amazon/amazon-ecs-agent:v1.0.0) with
+// tag that may be used to run the agent. It defaults to
+// 'amazon/amazon-ecs-agent:make', the version created locally by running
+// 'make'
+func RunAgent(t *testing.T, version *string) *TestAgent {
+	agent := &TestAgent{t: t}
+	agentImage := "amazon/amazon-ecs-agent:make"
+	if version != nil {
+		agentImage = *version
+	}
+	agent.Image = agentImage
+
+	dockerClient, err := docker.NewClient("unix:///var/run/docker.sock")
+	if err != nil {
+		t.Fatal(err)
+	}
+	agent.DockerClient = dockerClient
+
+	_, err = dockerClient.InspectImage(agentImage)
+	if err != nil {
+		err = dockerClient.PullImage(docker.PullImageOptions{Repository: agentImage}, docker.AuthConfiguration{})
+		if err != nil {
+			t.Fatal("Could not launch agent", err)
+		}
+	}
+	agentTempdir, err := ioutil.TempDir("", "ecs_integ_testdata")
+	if err != nil {
+		t.Fatal("Could not create temp dir for test")
+	}
+	logdir := filepath.Join(agentTempdir, "logs")
+	datadir := filepath.Join(agentTempdir, "data")
+	os.Mkdir(logdir, 0755)
+	os.Mkdir(datadir, 0755)
+	agent.TestDir = agentTempdir
+	t.Logf("Created directory %s to store test data in", agentTempdir)
+	err = agent.StartAgent()
+	if err != nil {
+		t.Fatal(err)
+	}
+	return agent
+}
+
+func (agent *TestAgent) StopAgent() error {
+	return agent.DockerClient.StopContainer(agent.DockerID, 10)
+}
+
+func (agent *TestAgent) StartAgent() error {
+	agent.t.Logf("Launching agent with image: %s\n", agent.Image)
+	logdir := filepath.Join(agent.TestDir, "logs")
+	datadir := filepath.Join(agent.TestDir, "data")
+	agentContainer, err := agent.DockerClient.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image: agent.Image,
+			ExposedPorts: map[docker.Port]struct{}{
+				"51678/tcp": struct{}{},
+			},
+			Env: []string{
+				"ECS_DATADIR=/data",
```

Comment:
Do we want a different cluster here?  Might make sense to create an `ecs-functional-tests` cluster and use that so there's no interference from other things running in the account.